### PR TITLE
feat(event_handler): add support for Pydantic models in Query and Header types

### LINKED ISSUE #5325: Feature request: improve support for parameter injection
### Why is this needed?

My apologies if there something already exists that addresses this issue. I searched all of the code and I just could not figure out a reusable way to handle parameter injection without rolling my own system. 

Request body and query string parsing as well as other aspects of request parsing are not reusable. For example, a pydantic based parameter injection system can be put in place at some point in the stack to make it more reusable. I've implemented this in a work project using a custom decorator

```python
from functools import wraps
from typing import Optional, Type
import inspect
import json

from aws_lambda_powertools.utilities.data_classes import APIGatewayProxyEvent
from pydantic import BaseModel, ValidationError

from lib.common.app import app

# NOTE: this utility does not map these types to the open api specification. 
def inject_params(
    query_model: Optional[Type[BaseModel]] = None,
    body_model: Optional[Type[BaseModel]] = None,
):
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            _ = args
            api_event: APIGatewayProxyEvent = app.current_event
            query_params = None
            body_params = None

            if query_model:
                query_dict = api_event.query_string_parameters or {}
                try:
                    query_params = query_model(**query_dict)
                except ValidationError as e:
                    # Raise validation error to be handled by existing middleware
                    raise e

            if body_model:
                try:
                    body_dict = json.loads(api_event.body or "{}")
                    body_params = body_model(**body_dict)
                except (ValidationError, json.JSONDecodeError) as e:
                    # Raise validation error to be handled by existing middleware
                    raise e

            # Dynamically inject only the parameters that the handler expects
            handler_sig = inspect.signature(func)
            if "query_params" in handler_sig.parameters and query_params is not None:
                kwargs["query_params"] = query_params
            if "body_params" in handler_sig.parameters and body_params is not None:
                kwargs["body_params"] = body_params

            return func(*args, **kwargs)

        return wrapper

    return decorator
```

Example usage:

```python

from http import HTTPStatus

from get_things import get_things
from lib.common.api_client.models.responses import PaginatedThingResponse
from lib.common.app import app, logger, metrics
from lib.common.aws import JsonResponse # custom class I implemented, ignore for the scope of this issue
from lib.common.dtos import GetThingsRequest
from lib.common.middleware.inject_params import inject_params
from lib.common.utils import run_async


@app.get("/thing")
@inject_params(query_model=GetThingsRequest)
def handler(
    query_params: GetThingsRequest,
) -> JsonResponse[PaginatedThingResponse]:
    response: PaginatedThingResponse = run_async(get_things, query_params)
    return JsonResponse(response, status_code=HTTPStatus.OK)
```

So this allows for reusable request parsing with pydantic.  The documented parsing strategy at this time is something that otherwise has to be repeated in every handler. Since this framework really leans into pydantic already I think this kind of functionality should exist. My implementation doesn't cover header parsing or anything else, but it could be extended to support it. 

I'd be happy to make an open source contribution if it makes sense. If so, where would be the best place? 

### Which area does this relate to?

_No response_

### Suggestion

_No response_

### Acknowledgment

- [x] This request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [x] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### LINKED ISSUE #7039: Feature request: alias generator for Query, Path or Header
### Use case

In a BaseModel you can setup an alias_generator.

It would be release nice to have something configurable for e.g. `Query`

### Solution/User Experience

Think about this:

```python
@app.get("/")
def list_something(max_results: Annotated[int, Query(alias="maxResults")]) -> Response[str]:
    ...
```
could become something like this:
```python
class CamelCaseQuery(Query):
    field_config = {"alias_generator": alias_generators.to_camel}

@app.get("/")
def list_something(max_results: Annotated[int, CamelCaseQuery()]) -> Response[str]:
    ...
```

### Alternative solutions

```markdown

```

### Acknowledgment

- [x] This feature request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by tonnico**
**Issue number:** closes #7039
closes #5325 

### Changes

You can use pydantic BaseModel for Query and Header.

### User experience

Grouping Query or Header params in one `BaseModel`.

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by anafalcao** [comment]
Hey @tonnico ! Thanks for opening this PR, we'll review it and get back to you

**Comment by leandrodamascena** [comment]
Hey @tonnico thanks a lot for working on this. I see we have some errors in the CI, can you fix them please and then we can start reviewing the code?

**Comment by tonnico** [comment]
I apologize to overlook `make pr`.

I still see some improvements and will fix the remaining issues.

**Comment by tonnico** [comment]
@leandrodamascena I fixed code smell and the remaining typing issues.

File: aws_lambda_powertools/event_handler/openapi/dependant.py [hunk 1]
```diff
@@ -275,7 +274,7 @@ def is_body_param(*, param_field: ModelField, is_path_param: bool) -> bool:
         return False
     elif is_scalar_field(field=param_field):
         return False
-    elif isinstance(param_field.field_info, (Query, Header)) and is_scalar_sequence_field(param_field):
+    elif isinstance(param_field.field_info, (Query, Header)):
```
Comment (tonnico) [inline aws_lambda_powertools/event_handler/openapi/dependant.py]: Maybe like this?

```suggestion
    elif isinstance(param_field.field_info, Param):
```

Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/openapi/dependant.py]: Hmm I think we need to make this specific. Because Form, File and Path, for example, they inherit from Param too and we don't support Pydantic models yet. I'm still reading your code and looks like we can easily add support for Pydantic models in those fields?

WDYT?

Comment (tonnico) [reply aws_lambda_powertools/event_handler/openapi/dependant.py]: Then `Param` should even better, because `Form`, `File` and `Path` are not `Body`?

```
def is_body_param
```

Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/openapi/dependant.py]: Ohhh ok, you're right. Both of them inherit from `FieldInfo` both those are different classes. Right stuff!

I'm reviewing the rest of the code.

**Comment by leandrodamascena** [comment]
Hi @tonnico! Just a heads up, I'm reading through the code and running several end-to-end tests to make sure everything works. It might take a while, but we'll definitely try to merge as soon as possible.

**Comment by tonnico** [comment]
Thanks for the update. Take your time. It's a big PR, and I'd be really sad if I broke something.

Thanks for taking the time to test everything.

**Comment by leandrodamascena** [comment]
Hey @tonnico I'm running another round of end2end tests with more advanced tests to see if it's not breakiung edge cases. Again, thanks for your patience and amazing work here.

**Comment by leandrodamascena** [comment]
Hi @tonnico! I've thoroughly tested this PR and haven't found any critical issues, except for some minor refactorings we need to make here. Honestly, this is great work. I still feel like we're not covering some use cases when working with Header/Query, but so far we haven't broken anything, which is great. I'm investing more time to write some additional tests and ensure we're ready for the merge - or work on some additional feedback.

We've had a few other PRs making changes to the OpenAPI utility, and I want to agree with you on the following strategy for merging this PR:

1/ I've merged this PR https://github.com/aws-powertools/powertools-lambda-python/pull/7227 - and it will be included in the next release on September 25th.

2/ We can work to merge **THIS PR** sometime after September 25th and include it in the release on October 7th.

3/ I'm adding the "on hold" label just to make it clear to everyone else.

Thank you very much for your work, patience, and time invested here.

Btw, I see there is a conflict in the tests files, but it's easy to fix.

**Comment by tonnico** [comment]
Thanks @leandrodamascena for the good starting point from your initial PR.

I just fixed the merge conflict.

**Comment by leandrodamascena** [comment]
Reviewing this now.

**Review by leandrodamascena — APPROVED** [review]
Hi @tonnico! Thank you so much for your patience while waiting for my final review. I went through 4 or 5 rounds of review. I tried to create a scenario where I could break the code/functionality, and fortunately, I couldn't. I ran e2e tests with this in different scenarios.

This code is complex. You've made some important refactorings, and I can think of two or three different ways to do this implementation with the current codebase. But, to be honest, I don't know if I could do it better than you did here. The abstraction you create for some functions, the refactoring to reduce complexity, and the clarity of the code are very good. I hope I'm not missing any major potential problem I haven't seen, but our current tests are passing and that brings more peace.

That's really a super PR, thanks a lot. I sent a commit with some changes in the doc and we are ready to merge tomorrow or Friday.

THANKS A LOT! APPROVED!!!!!

**Comment by tonnico** [comment]
Wow, thanks so much for your review and the nice feedback! Appreciate all the time you spent testing and checking different scenarios. Glad it held up. Looking forward to the merge!

**Review by sthulb — APPROVED** [review]
